lab-dustin/server.js

```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+const net = require('net');
+const PORT = process.env.PORT || 3000;
+const server = net.createServer();
+const constructor = require('./client-constructor');
+
+const EE = require('events');
+const ee = new EE();
+
+const pool = [];
+
+ee.on('@all',function(sender,message){
+  pool.forEach(receiver => {
+    if (receiver.id === sender.id) return;
+
+    receiver.socket.write(`${sender.id}: ${message}\r\n`);
+  });
+});
+
+ee.on('@username',function(sender,name){
+  console.log(name,sender.id);
+  var usernameExists = false;
+  pool.forEach(function(user){
+    if (user.id === name){
+      sender.socket.write('Username already exists.\r\n');
+      usernameExists = true;
+    }
+  });
+
+  if (!usernameExists){
+    sender.socket.write(`${sender.id} changed username to: ${name}\r\n`);
+    sender.id = name;
+  }
+});
+
+ee.on('@dm',function(sender,destination,message){
+  pool.forEach(function(receiver){
+    console.log(receiver);
+    if (receiver.id === destination){
+      receiver.socket.write(`${sender.id}: ${message}\r\n`);
+    }
+  });
+});
+
+server.on('connection',function(socket){
+  var client = new constructor.Client(socket);
+  pool.push(client);
+  console.log(`${client.id} joined. Users active: ${pool.length}`);
+  socket.write('Commands are prefixed with @\r\n');
+
+  socket.on('data',function(data){
+    var startCommand = data.toString('utf-8').indexOf('@');
+    var endCommand = data.toString('utf-8').indexOf(' ');
+    var command = data.toString('utf-8').slice(startCommand,endCommand);
+
+    if (command === '@all'){
+      ee.emit('@all',client,data.toString().trim().slice(++endCommand));
```

Comment:
`++endCommand` here feels a little strange, since you're not using the `endCommand` value again. I'd probably go with `endCommand+1` to make that more clear.

(And note that this is a spot where you'd need to have used `++` first so that the result is the incremented value.)